edn/src/matcher.rs

```diff
@@ -0,0 +1,541 @@
+
+    fn matches_placeholder(pattern: &'a Value) -> Option<(&'a String)> {
+        match *pattern {
+            Value::PlainSymbol(symbols::PlainSymbol(ref s)) => if s.starts_with('?') { Some(s) } else { None },
+            _ => None
+        }
+    }
+}
+
+/// Pattern matcher for EDN values utilizing specified pattern matching rules.
+/// For example, using this with `DefaultPatternMatchingRules`:
+/// * `[_]` matches an arbitrary one-element vector;
+/// * `[_ _]` matches an arbitrary two-element vector;
+/// * `[?x ?x]` matches `[1 1]` and `[#{} #{}]` but not `[1 2]` or `[[] #{}]`;
+struct Matcher<'a> {
+    placeholders: RefCell<HashMap<&'a String, &'a Value>>
+}
+
+impl<'a> Matcher<'a> {
+    /// Creates a Matcher instance.
+    fn new() -> Matcher<'a> {
+        Matcher {
+            placeholders: RefCell::default()
+        }
+    }
+
+    /// Performs pattern matching between two EDN `Value` instances (`value`
+    /// and `pattern`) utilizing a specified pattern matching ruleset `T`.
+    /// Returns true if matching succeeds.
+    pub fn match_with_rules<T>(value: &'a Value, pattern: &'a Value) -> bool
+    where T: PatternMatchingRules<'a, Value> {
+        let matcher = Matcher::new();
+        matcher.match_internal::<T>(value, pattern)
+    }
+
+    /// Recursively traverses two EDN `Value` instances (`value` and `pattern`)
+    /// performing pattern matching. Note that the internal `placeholders` cache
+    /// may not be empty on invocation.
+    fn match_internal<T>(&self, value: &'a Value, pattern: &'a Value) -> bool
+    where T: PatternMatchingRules<'a, Value> {
+        use Value::*;
+
+        if T::matches_any(pattern) {
+            true
+        } else if let Some(symbol) = T::matches_placeholder(pattern) {
+            let mut cache = self.placeholders.borrow_mut();
+            if !cache.contains_key(symbol) {
+                cache.insert(symbol, value); true
+            } else {
+                value == *cache.get(symbol).unwrap()
+            }
+        } else {
+            match (value, pattern) {
+                (&Vector(ref a), &Vector(ref b)) =>
+                    diff_with(a, b, |a, b| self.match_internal::<T>(a, b)).is_none(),
+                (&List(ref a), &List(ref b)) =>
+                    diff_with(a, b, |a, b| self.match_internal::<T>(a, b)).is_none(),
+                (&Set(ref a), &Set(ref b)) =>
+                    a.len() == b.len() &&
+                    a.iter().all(|av| b.iter().any(|bv| self.match_internal::<T>(av, bv))) &&
+                    b.iter().all(|bv| a.iter().any(|av| self.match_internal::<T>(av, bv))),
+                (&Map(ref a), &Map(ref b)) =>
+                    a.len() == b.len() &&
+                    a.iter().all(|(ak, av)| b.iter().any(|(bk, bv)| self.match_internal::<T>(ak, bk) && self.match_internal::<T>(av, bv))) &&
+                    b.iter().all(|(bk, bv)| a.iter().any(|(ak, av)| self.match_internal::<T>(ak, bk) && self.match_internal::<T>(av, bv))),
```

Comment:
This ended up being just a tad bit more complicated than expected, since we can't just compare collections directly with equality since that would render the patterns useless. For example, `[_]` would never be equal to `[1]` unless we mess with the `Eq` trait implementation instead of deriving the default.